Common/src/linear_algebra/CSysMatrix.cpp

```diff
@@ -815,136 +708,152 @@ void CSysMatrix<ScalarType>::ComputeJacobiPreconditioner(const CSysVector<Scalar
 template<class ScalarType>
 void CSysMatrix<ScalarType>::BuildILUPreconditioner(bool transposed) {
 
-  unsigned long index, index_, iVar;
-  ScalarType *Block_ij;
-  const ScalarType *Block_jk;
-  long iPoint, jPoint, kPoint;
-
-  /*--- Copy block matrix, note that the original matrix
-   is modified by the algorithm, so that we have the factorization stored
-   in the ILUMatrix at the end of this preprocessing. ---*/
-
-  for (iVar = 0; iVar < nnz_ilu*nVar*nEqn; iVar++) ILU_matrix[iVar] = 0.0;
-
-  for (iPoint = 0; iPoint < (long)nPointDomain; iPoint++) {
-    for (index = row_ptr[iPoint]; index < row_ptr[iPoint+1]; index++) {
-      jPoint = col_ind[index];
-      if (transposed) {
-        Block_ij = GetBlock(jPoint, iPoint);
-        SetBlockTransposed_ILUMatrix(iPoint, jPoint, Block_ij);
-      } else {
-        Block_ij = GetBlock(iPoint, jPoint);
-        SetBlock_ILUMatrix(iPoint, jPoint, Block_ij);
+  /*--- Copy block matrix to compute factorization in-place. ---*/
+
+  if ((ilu_fill_in == 0) && !transposed) {
+    /*--- ILU0, direct copy. ---*/
+    SU2_OMP_PAR_FOR_STAT(OMP_STAT_SIZE)
+    for (auto iVar = 0ul; iVar < nnz*nVar*nVar; ++iVar)
+      ILU_matrix[iVar] = matrix[iVar];
+  }
+  else {
+    /*--- ILUn clear the ILU matrix first, for ILU0^T
+     *    the copy takes care of the clearing. ---*/
+    if (ilu_fill_in > 0) {
+      SU2_OMP_PAR_FOR_STAT(OMP_STAT_SIZE)
+      for (auto iVar = 0ul; iVar < nnz_ilu*nVar*nVar; iVar++)
+        ILU_matrix[iVar] = 0.0;
+    }
+
+    /*--- Transposed or ILUn, traverse matrix to access its blocks
+     *    sequentially and set them in the ILU matrix. ---*/
+    SU2_OMP_PAR_FOR_DYN(omp_chunk_size)
+    for (auto iPoint = 0ul; iPoint < nPointDomain; iPoint++) {
+      for (auto index = row_ptr[iPoint]; index < row_ptr[iPoint+1]; index++) {
+        auto jPoint = col_ind[index];
+        if (transposed) {
+          SetBlockTransposed_ILUMatrix(jPoint, iPoint, &matrix[index*nVar*nVar]);
+        } else {
+          SetBlock_ILUMatrix(iPoint, jPoint, &matrix[index*nVar*nVar]);
+        }
       }
     }
   }
 
   /*--- Transform system in Upper Matrix ---*/
 
-  for (iPoint = 1; iPoint < (long)nPointDomain; iPoint++) {
+  SU2_OMP_PARALLEL_ON(omp_num_parts)
+  {
+  int thread = omp_get_thread_num();
+  const auto begin = omp_partitions[thread];
+  const auto end = omp_partitions[thread+1];
+
+  ScalarType weight[MAXNVAR*MAXNVAR], aux_block[MAXNVAR*MAXNVAR];
+
+  for (auto iPoint = begin+1; iPoint < end; iPoint++) {
```

Comment:
For preconditioners that are serial by nature we have a parallel section where each thread works on its own large chunk of the matrix.